packages/fx-core/src/component/utils/fileOperation.ts

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import * as fs from "fs-extra";
+import klaw from "klaw";
+import AdmZip from "adm-zip";
+import ignore, { Ignore } from "ignore";
+import path from "path";
+import { DeployConstant } from "../constant/deployConstant";
+
+/**
+ * Asynchronously zip a folder and return buffer
+ * @param sourceDir base dir
+ * @param notIncluded block list
+ * @param cache zip cache file location
+ */
+export async function zipFolderAsync(
+  sourceDir: string,
+  cache: string,
+  notIncluded: Ignore
+): Promise<Buffer> {
+  const normalizeTime = (t: number) => Math.floor(t / DeployConstant.ZIP_TIME_MS_GRANULARITY);
+
+  const tasks: Promise<void>[] = [];
+  const zipFiles = new Set<string>();
+  const ig = notIncluded ?? ignore();
+  const zip = (await readZip(cache)) || new AdmZip();
+
+  const addFileIntoZip = async (
+    zp: AdmZip,
+    filePath: string,
+    zipPath: string,
+    stats?: fs.Stats
+  ) => {
+    const content = await fs.readFile(filePath);
+    zp.addFile(zipPath, content);
+    if (stats) {
+      (zp.getEntry(zipPath)!.header as any).time = stats.mtime;
+    }
+  };
+
+  await forEachFileAndDir(sourceDir, (itemPath: string, stats: fs.Stats) => {
+    const relativePath: string = path.relative(sourceDir, itemPath);
+    if (relativePath && !stats.isDirectory() && ig.filter([relativePath]).length > 0) {
+      zipFiles.add(relativePath);
+
+      const entry = zip.getEntry(relativePath);
+      if (entry) {
+        // The header is an object, the ts declare of adm-zip is wrong.
+        const header = entry.header as any;
+        const mtime = header && header.time;
+        // Some files' mtime in node_modules are too old, which may be invalid,
+        // so we arbitrarily add a limitation to update this kind of files.
+        // If mtime is valid and the two mtime is same in two-seconds, we think the two are same file.
+        if (
```

Comment:
I am not sure if we still need this logic for incremental zip optimization?

Reply:
Maybe we can remove the last trust time logic  and keep the last two-seconds logic